compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -1180,6 +1180,21 @@ pub fn fn_can_unwind(tcx: TyCtxt<'_>, fn_def_id: Option<DefId>, abi: SpecAbi) ->
                 return false;
             }
         }
+
+        // Functions in compiler builtins can never unwind.
+        //
+        // The `compiler-builtins` crate exports a bunch of LLVM intrinsics which are defined
+        // with the C ABI, so it must not unwind. If such functions call into unwindable
+        // functions, rustc will inject unwind guards to prevent unwind leaking out. This is
+        // normally okay, but for `compiler-builtins` crate it has the additional requirement to
+        // not call into libcore, so we *must not* generate such guards. One way to do it is to
+        // ensure those exported functions never call into unwindable functions, so we make
+        // every single function defined in compiler builtins nounwind.
+        //
+        // See rust-lang/compiler-builtins#578 for context.
+        if tcx.is_compiler_builtins(did.krate) && !tcx.is_foreign_item(did) {
+            return false;
+        }
```

Comment:
FWIW, if this works I think it should be landed as a separate change, just in case it causes problems. The stabilization PR should only do stabilization and nothing else, ideally.

Reply:
The issue is that since compiler_builtins crate currently don't have `c_unwind` feature enabled, landing this change separately is not really meaningful. Without stabilisation PR, we'll need to manual enable `c_unwind` feature in `compiler_builtins` for this to be effective, but doing so is arguably a bad idea since it breaks the crate for all older compiler versions.

Reply:
> it breaks the crate for all older compiler versions.

I don't think that matters, the crate is very tied to latest nightly rustc.

Ideally the crate can opt-in to the new feature before it is stabilized, so that we can check if it actually works before landing the final PR.

Reply:
> The issue is that since compiler_builtins crate currently don't have c_unwind feature enabled

Given that compiler_builtins is a critical crate when it comes to unwinding, and given that just enabling the feature gate has a side-effect (which is somewhat unconventional), I think that there absolutely should be a PR to compiler_builtins enabled this feature before we do a stabilization PR. That way we can explore what "stabilize this only for compiler_builtins" looks like. Everything else seems way too risky.

Reply:
Submitted https://github.com/rust-lang/compiler-builtins/pull/583 to do that.